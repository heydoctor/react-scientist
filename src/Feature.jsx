// @flow

import React, { ElementType, Fragment } from 'react';
import Cookie from 'js-cookie';

type VariableValuesObject = {
  [key: string]: boolean | number | string | null,
};

type ChildrenRenderFunction = (
  isEnabled: boolean,
  variables: VariableValuesObject
) => ElementType;

type FeatureProps = {
  children: ChildrenRenderFunction,
  disableTracking?: boolean,
  domain?: string,
  enabledPercent?: number, //value from 0 to 100 indicate probability of enabling this feature to user
  isEnabled?: boolean, // control enabling feature from outside
  name: string,
  userId?: string,
  variables: VariableValuesObject,
};

type StartArgs = {
  featureName: string,
  isEnabled: boolean,
  variables: number,
};

const throwError = message => {
  throw new Error(`<Feature />: ${message}`);
};

/**
 * Feature component allows to evaluate the state of a feature flag and its variables for a given use
 * Can be controlled from outside by isEnabled flag,
 * or internally calculate probability of enabling for user and store value in cookies
 *
 * @param props
 * @returns {JSX.Element}
 * @constructor
 */
const Feature = ({
  children,
  disableTracking,
  domain,
  enabledPercent,
  isEnabled,
  userId,
  name,
  variables,
}: FeatureProps) => {
  const storeKey = `scientist.feature.${
    typeof userId === 'number' || typeof userId === 'string' ? userId : 'guest'
  }.${name}`;

  const storedValue = Cookie.get(storeKey);

  //if no cookie and no props from parent we need to calculate probability of enabling feature
  let isEnabledCalculated = isEnabled;
  if (typeof storedValue === 'undefined' && typeof isEnabled === 'undefined') {
    const chance = Math.random() * 100; //range from 0 to 100
    isEnabledCalculated = enabledPercent > chance;
  }
  //if no cookie, save value and report it
  if (typeof storedValue === 'undefined') {
    Cookie.set(storeKey, isEnabled, { domain });
    if (!disableTracking) {
      Feature.onStart({
        featureName: name,
        isEnabled,
        variables,
      });
    }
  }

  return <Fragment>{children(isEnabledCalculated, variables)}</Fragment>;
};

Feature.onStart = (args: StartArgs) => {
  throwError(
    'This method should be overwritten somewhere before React is mounted, e.g. Feature.onStart = () => {...}'
  );
};

export default Feature;
