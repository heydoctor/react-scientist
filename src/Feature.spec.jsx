import React from 'react';
import { render, cleanup } from 'react-testing-library';
import Cookie from 'js-cookie';
import Feature from './Feature';

const children = (isEnabled: boolean, variables: VariableValuesObject) =>
  isEnabled && <span>{JSON.stringify(variables)}</span>;

describe('<Feature />', () => {
  beforeAll(() => {
    Feature.onStart = jest.fn();
  });

  afterEach(() => {
    Feature.onStart.mockReset();
  });

  test('storeKey is set to guest if no userId is provided', () => {
    const name = 'Test';
    const { container } = render(
      <Feature name={name} enabledPercent={100}>
        {children}
      </Feature>
    );
    expect(Cookie.get(`scientist.feature.guest.${name}`)).toBeDefined();
  });

  test('storeKey uses userId if provided', () => {
    const name = 'Test';
    const userId = 3;
    const { container } = render(
      <Feature name={name} enabledPercent={100} userId={userId}>
        {children}
      </Feature>
    );
    expect(Cookie.get(`scientist.feature.${userId}.${name}`)).toBeDefined();
  });

  test('calls Feature.onStart only once', () => {
    const name = 'onStart';
    const { rerender } = render(
      <Feature name={name} enabledPercent={100}>
        {children}
      </Feature>
    );
    expect(Feature.onStart).toHaveBeenCalledTimes(1);
    rerender(
      <Feature name={name} enabledPercent={100}>
        {children}
      </Feature>
    );
    expect(Feature.onStart).toHaveBeenCalledTimes(1);
  });

  test('renders variables by enabledPercent', () => {
    const name = 'Render';
    const variables = { key: 'testVariables' };
    const { container } = render(
      <Feature name={name} enabledPercent={100} variables={variables}>
        {children}
      </Feature>
    );

    expect(JSON.stringify(variables)).toContain(container.textContent);
  });

  test('renders variables by enabled props', () => {
    const name = 'Render';
    const variables = { key: 'testVariables' };
    const { container } = render(
      <Feature name={name} isEnabled={true} variables={variables}>
        {children}
      </Feature>
    );

    expect(JSON.stringify(variables)).toContain(container.textContent);
  });

  test('renders null if not enabled', () => {
    const name = 'Null';
    const variables = { key: 'testVariables' };
    const { container } = render(
      <Feature name={name} enabledPercent={0} variables={variables}>
        {children}
      </Feature>
    );

    expect(container.textContent).toContain('');
  });
});
