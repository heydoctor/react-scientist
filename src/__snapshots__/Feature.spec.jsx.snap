// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature#onStart throws 1`] = `"<Feature />: This method should be overwritten somewhere before React is mounted, e.g. Feature.onStart = () => {...}"`;
